app/database/itemStore.go

```diff
@@ -38,9 +38,31 @@ func (s *ItemStore) Insert(data *Item) error {
 	return s.db.insert(s.tableName(), data)
 }
 
+// All creates a composable query without filtering
+func (s *ItemStore) All() *DB {
+	return &DB{s.db.Table(s.tableName())}
+}
+
 // IsValidHierarchy gets an ordered set of item ids and returns whether they forms a valid item hierarchy path from a root
 func (s *ItemStore) IsValidHierarchy(ids []int64) (bool, error) {
-	return false, nil
+	l := len(ids) - 1
+	if l < 1 {
+		return false, fmt.Errorf("not enough ids %v", ids)
+	}
+
+	items := s.Items().All().Where("items.ID IN (?)", ids[1:]).Or("items.ID = ? AND items.sType = ?", ids[0], "Root")
```

Comment:
Using the `items` table is only needed to check the root (trivial request). The rest of the filter (with the expensive `IN (...)` is useless as its `items_items` which is checking the relations. 
Actually I think I see what you tried to do, by doing everything in a single query instead of two, but in this case, this the very inefficient I think. Checking the root in one first request is so trivial. And even with a single request, doing it with a basic join (without any condition on `items`) and then checking the root in code would be better (in my opinion).